sealtk/core/KwiverTrackModel.cpp

```diff
@@ -0,0 +1,236 @@
+  : AbstractItemModel{parent}, d_ptr{new KwiverTrackModelData}
+{
+}
+
+// ----------------------------------------------------------------------------
+KwiverTrackModel::KwiverTrackModel(
+  QSharedDataPointer<KwiverTrackModelData> const& d, QObject* parent)
+  : AbstractItemModel{parent}, d_ptr{d}
+{
+}
+
+// ----------------------------------------------------------------------------
+KwiverTrackModel::~KwiverTrackModel()
+{
+}
+
+// ----------------------------------------------------------------------------
+int KwiverTrackModel::rowCount(QModelIndex const& parent) const
+{
+  QTE_D_SHARED();
+
+  if (parent.isValid())
+  {
+    // TODO detections?
+    return 0;
+  }
+
+  return static_cast<int>(d->tracks.size());
+}
+
+// ----------------------------------------------------------------------------
+QVariant KwiverTrackModel::data(QModelIndex const& index, int role) const
+{
+  // TODO detections?
+  if (this->checkIndex(index, IndexIsValid | ParentIsInvalid))
+  {
+    QTE_D_SHARED();
+
+    auto const& track = d->tracks[static_cast<unsigned>(index.row())];
+    switch (role)
+    {
+      case core::NameRole:
+      case core::LogicalIdentityRole:
+        return static_cast<qint64>(track.track->id());
+
+      case core::StartTimeRole:
+        if (!track.track->empty())
+        {
+          auto const& s =
+            std::static_pointer_cast<kv::object_track_state>(
+              track.track->front());
+          return QVariant::fromValue(s->time());
+        }
+        return {};
+
+      case core::EndTimeRole:
+        if (!track.track->empty())
+        {
+          auto const& s =
+            std::static_pointer_cast<kv::object_track_state>(
+              track.track->back());
+          return QVariant::fromValue(s->time());
+        }
+        return {};
+
+      case core::UserVisibilityRole:
+        return track.visible;
+
+      default:
+        break;
+    }
+  }
+
+  return AbstractItemModel::data(index, role);
+}
+
+// ----------------------------------------------------------------------------
+void KwiverTrackModel::addTracks(
+  kv::object_track_set_sptr const& trackSet)
+{
+  QVector<kv::track_sptr> newTracks;
+
+  if (trackSet)
+  {
+    auto const& tracks = trackSet->tracks();
+    for (auto&& track : tracks | kvr::transform(cleanTrack) | kvr::valid)
+    {
+      newTracks.append(std::move(track));
+    }
+  }
+
+  if (!newTracks.isEmpty())
+  {
+    QTE_D_DETACH();
+
+    auto const oldRows = this->rowCount({});
+    auto const newRows = newTracks.size();
+
+    this->beginInsertRows({}, oldRows, newRows);
+
+    for (auto& track : newTracks)
+    {
+      d->tracks.emplace_back(std::move(track));
+    }
+
+    this->endInsertRows();
+  }
+}
+
+// ----------------------------------------------------------------------------
+void KwiverTrackModel::setTracks(
+  kv::object_track_set_sptr const& trackSet)
+{
+  this->beginResetModel();
+
+  with_expr(qtScopedBlockSignals{this})
```

Comment:
`with_expr` causes build errors on my machine.

Reply:
Oh? What errors?

ViViA has been using it for some time...